### DISCUSSION TITLE
New renewable site clustering pipeline

### DISCUSSION THREAD

**Discussion by gschivley** [PowerGenome/PowerGenome#216] [Ideas]
Hi all, I'm looking for some feedback on a big planned feature change.

To date, PowerGenome has relied on pre-clustered wind and solar sites.[^1]  The drawback has been a lack of flexibility in clustering methods. Moving forward I want to shift to a method where _users will store all the renewable site details and generation profiles locally_ and PG can flexibly combine a number of different methods to cluster them.

Moving to a flexible clustering pipeline that can be specified in YAML and accommodate new methods or different features is challenging. I've started outlining a system and want to get feedback on both the overall system and specific clustering algorithms that might be useful to include.

The process below assumes that users start with a list of potential project sites (i.e. the [Candidate Project Areas](https://zenodo.org/record/5021146#.YweMnC-B3Fw) developed at Princeton). Each site will have a generation profile and various numeric/categorical metadata.[^2]

**Step 1: Filter**
Filtering can be based on numeric features (LCOE < $150/MWh), categorical (not on prime farmland), or geospatial (exclude areas in a shapefile provided by the user). The goal here is to reduce the number of sites that will be included in the final clusters, either by removing sites that should not be allowed or removing sites that are unlikely to be built for financial reasons.

**Step 2: Group**
Within each model region, are there any geographic/categorical features that should be used to split sites? This could be by county, IPM region, prime farmland vs not, BOEM vs non-BOEM offshore lease areas, etc. Sites will be split into groups and the next clustering steps will be applied within each group.

**Step 3: Bin**
Within each group, are there any numeric features that should be used to bin into sub-groups? Examples could be interconnection cost, LCOE, population density, etc. This step might be considered a variant of Step 4 (cluster). Binning could be either equal width ([pandas cut](https://pandas.pydata.org/docs/reference/api/pandas.cut.html)) or equal size ([pandas qcut](https://pandas.pydata.org/docs/reference/api/pandas.qcut.html#pandas.qcut)). Users would need to provide the explicit number of bins, the average capacity per bin (e.g. 1 bin for every 10 GW), the bin edges, or specific quantiles.

**Step 4: Cluster**
Within each sub-group, apply a specific clustering algorithm using one or more specified features. This could be a hierarchical method using hourly generation profiles or k-means using lat/lon. Users would need to provide the explicit number of clusters or the average capacity per cluster (e.g. 1 cluster for every 2 GW).

The idea is that these steps form a pipeline. Any of the steps above can be skipped or applied multiple times. Steps can be different in each model region. Want to use all sites and group by county in one region? Go for it. In another region you can filter by LCOE, quartile bin by interconnection cost, and cluster by generation profile. Or apply the same pipeline to all regions.

I'm looking for feedback on:
1. The general structure. Does this pipeline seem reasonable? Is there anything it's missing? PG should be able to accept new clustering methods in Step 4 but is there anything that wouldn't fit in the general structure?
2. Implementation. Do you know of any examples where something like this has been done? I'll have a nested dictionary of strings that will need to select functions, serve as arguments, and generally make the whole pipeline work.
3. Specific clustering algorithms that should be included in Step 4. Generic scikit/sklearn k-means and hierarchical clustering are obvious choices. There are also packages such as [tslearn](https://tslearn.readthedocs.io/en/stable/user_guide/clustering.html) that could be used. 

[^1]: As a way to a) reduce required file sizes, b) speed up code, and c) maintain flexibility in the model regions/capacity per region/number of clusters per region.
[^2]: Metadata such as the location (lat/lon, county, state, region, etc), capacity, interconnection cost, generation profile, LCOE (calculated using an assumed resource cost), population density, human modification index of the land, or if the site is located on prime farmland.

**Comment by zaneselvans** [comment]
@bendnorman do you have suggestions from Dagster here? This seems like the kind of thing it's good at coordinating.

**Reply by bendnorman** [reply]
Hello! Yes, [Dagster](https://github.com/dagster-io/dagster) is an orchestration framework that allows you to parameterize data pipelines using [`ConfigSchemas`](https://docs.dagster.io/concepts/configuration/config-schema). Each input to a node in the graph is either an output of another node or a parameter specified in config. 

You can also [programmatically construct DAGs](https://docs.dagster.io/concepts/ops-jobs-graphs/graphs#constructing-dependencies) laid out in YAML though I haven't tried. 

We are currently integrating [Dagster into PUDL](https://github.com/catalyst-cooperative/pudl/pull/1835) if you want a familiar example though it is a work in progress.

**Reply by gschivley** [reply]
Thanks, will check it out!

**Comment by gschivley** [comment]
I _think_ that the YAML to code pipeline can be handled with a dictionary that matches the string names of steps to functions.

```py
def plus(a, b):
  return a + b

def sub(a, b):
  return a - b

def double(a):
  return a * 2

d = {
    "plus": plus,
    "sub": sub,
    "double": double,
}

```

Then steps can be defined in YAML and piped together 

```py
import pandas as pd
import yaml

yml = """
first: 
  fn: plus
  params: [3]
second:
  fn: sub
  params: [5]
third:
  fn: double
  params: []
"""
s = yaml.safe_load(yml)

df = pd.Series([1, 2, 3])

df.pipe(
    d[s["first"]["fn"]], *s["first"].get("params", [])
).pipe(
    d[s["second"]["fn"]], *s["second"].get("params", [])
).pipe(
    d[s["third"]["fn"]], *s["third"].get("params", [])
)
```

Has anyone seen something like this before or implemented it themselves? An example or two that I can reference would be helpful!

**Comment by bradenpecora** [comment]
Have you considered incorporating a decomposition step such as [PCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html)? Could be particularly useful for clustering with generation profiles.

**Reply by gschivley** [reply]
Have not thought of it. Would it fit into the framework I outlined above? Can you give an example of how you'd use it?

**Reply by bradenpecora** [reply]
Suppose the matrix/DataFrame X contains 8760 generation profiles that we would like to cluster on. Rows are timepoints and columns are each respective profile. Thus, X is 8760 rows by the number of profiles. We could reduce the number of rows clustered on to a number of our choosing:

```python
import numpy as np
from sklearn.decomposition import PCA

X = ...
pca = PCA(n_components=5) # 5 is the user choice
out = np.transpose(
      pca.fit_transform(np.transpose(X))
)
```
`out` will be a 5 rows by number of profiles matrix. The time domain has been reduced and is simpler to cluster with.

This would be easy to incorporate into the pipeline before any clustering step. The main practical use would be for time domain reduction. If you wanted to cluster on a dozen different variables (not necessarily time series) that may be related, you could cluster with a smaller set of "representative" variables.

The main goal would be to speed up computation. It wouldn't be necessary though. Maybe a "v1.1"

**Comment by joshdr83** [comment]
I have thoughts, but am swamped by a deadline for mid next week. When are you wanting to have a path forward by?

**Reply by gschivley** [reply]
This is still in the early-ish stage. I expect it will be implemented over the next couple months, although early versions could be set up if you guys have something specific you want to use.
